File: src/main/java/com/paragon/client/systems/module/impl/render/Breadcrumbs.java
package com.paragon.client.systems.module.impl.render;

import com.paragon.api.util.render.ColourUtil;
import com.paragon.client.systems.module.Module;
import com.paragon.client.systems.module.Category;
import com.paragon.client.systems.module.setting.Setting;
import net.minecraft.util.math.Vec3d;

import java.awt.*;
import java.util.LinkedList;

import static org.lwjgl.opengl.GL11.*;

/**
 * @author Wolfsurge
 */
public class Breadcrumbs extends Module {

    public static Breadcrumbs INSTANCE;

    // Settings
    public static Setting<Boolean> infinite = new Setting<>("Infinite", false)
            .setDescription("Breadcrumbs last forever");

    public static Setting<Float> lifespanValue = new Setting<>("Lifespan", 100f, 1f, 1000f, 1f)
            .setDescription("The lifespan of the positions in ticks")
            .setVisibility(() -> !infinite.getValue());

    public static Setting<Float> lineWidth = new Setting<>("LineWidth", 1f, 0.1f, 5f, 0.1f)
            .setDescription("The width of the lines");

    public static Setting<Color> colour = new Setting<>("Colour", new Color(185, 17, 255))
            .setDescription("The colour of the breadcrumbs");

    public static Setting<Boolean> rainbow = new Setting<>("Rainbow", true)
            .setDescription("Makes the trail a rainbow");

    private final LinkedList<Position> positions = new LinkedList<>();
    private int colourHue = 0;

    public Breadcrumbs() {
        super("Breadcrumbs", Category.RENDER, "Draws a trail behind you");

        INSTANCE = this;
    }

    @Override
    public void onDisable() {
        // Clear positions when we disable
        positions.clear();
    }

    @Override
    public void onTick() {
        if (nullCheck() || mc.player.ticksExisted <= 20) {
            // We may have just loaded into a world, so we need to clear the positions
            positions.clear();
            return;
        }

        // Create position
        Position pos = new Position(new Vec3d(mc.player.lastTickPosX, mc.player.lastTickPosY, mc.player.lastTickPosZ), new Color(ColourUtil.getRainbow(4, 1, colourHue)));
        colourHue++;

        // Add position
        positions.add(pos);

        // Update positions
        positions.forEach(Position::update);

        // Remove old positions
        positions.removeIf(p -> !p.isAlive() && !infinite.getValue());
    }

    @Override
    public void onRender3D() {
        glPushMatrix();

        // GL stuff
        glDisable(GL_TEXTURE_2D);
        glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
        glEnable(GL_LINE_SMOOTH);
        glEnable(GL_BLEND);
        glDisable(GL_DEPTH_TEST);
        glLineWidth(lineWidth.getValue());

        // Disable lighting
        mc.entityRenderer.disableLightmap();

        glBegin(GL_LINE_STRIP);

        for (Position pos : positions) {
            double renderPosX = mc.getRenderManager().viewerPosX;
            double renderPosY = mc.getRenderManager().viewerPosY;
            double renderPosZ = mc.getRenderManager().viewerPosZ;

            ColourUtil.setColour(rainbow.getValue() ? pos.getColour().getRGB() : colour.getValue().getRGB());
            glVertex3d(pos.getPosition().x - renderPosX, pos.getPosition().y - renderPosY, pos.getPosition().z - renderPosZ);
        }

        // Reset colour
        glColor4d(1, 1, 1, 1);

        // End GL
        glEnd();
        glEnable(GL_DEPTH_TEST);
        glDisable(GL_LINE_SMOOTH);
        glDisable(GL_BLEND);
        glEnable(GL_TEXTURE_2D);
        glPopMatrix();
    }

    class Position {
        // Vec3d of position
        private final Vec3d position;
        // Position's colour
        private final Color colour;
        // Position's lifespan
        private long lifespan = lifespanValue.getValue().longValue();

        public Position(Vec3d position, Color colour) {
            this.position = position;
            this.colour = colour;
        }

        /**
         * Decreases the lifespan of the position
         */
        public void update() {
            lifespan--;
        }

        /**
         * Checks if the position is alive
         *
         * @return If the position is alive
         */
        public boolean isAlive() {
            return lifespan > 0;
        }

        /**
         * Gets the position
         *
         * @return The position
         */
        public Vec3d getPosition() {
            return position;
        }

        /**
         * Gets the position's colour
         *
         * @return The position's colour
         */
        public Color getColour() {
            return colour;
        }
    }
}

File: src/main/kotlin/com/paragon/impl/module/render/Breadcrumbs.kt
package com.paragon.impl.module.render

import com.paragon.impl.module.Module
import com.paragon.impl.setting.Setting
import com.paragon.util.player.PlayerUtil
import com.paragon.util.render.ColourUtil.setColour
import com.paragon.impl.module.Category
import com.paragon.util.anyNull
import net.minecraft.util.math.Vec3d
import org.lwjgl.opengl.GL11.*
import java.awt.Color
import java.util.*

/**
 * @author Surge
 */
object Breadcrumbs : Module("Breadcrumbs", Category.RENDER, "Draws a trail behind you") {

    private val infinite = Setting(
        "Infinite", false
    ) describedBy "Breadcrumbs last forever"

    private val lifespanValue = Setting(
        "Lifespan", 100f, 1f, 1000f, 1f
    ) describedBy "The lifespan of the positions in ticks" visibleWhen { !infinite.value }

    private val lineWidth = Setting(
        "LineWidth", 1f, 0.1f, 5f, 0.1f
    ) describedBy "The width of the lines"

    private val colour = Setting(
        "Colour", Color(185, 17, 255)
    ) describedBy "The colour of the breadcrumbs"

    private val rainbow = Setting(
        "Rainbow", true
    ) describedBy "Makes the trail a rainbow"

    private val positions = LinkedList<Position>()
    private var colourHue = 0

    override fun onDisable() {
        // Clear positions when we disable
        positions.clear()
    }

    override fun onTick() {
        if (minecraft.anyNull || minecraft.player.ticksExisted <= 20) {
            // We may have just loaded into a world, so we need to clear the positions
            positions.clear()
            return
        }

        // Create position
        val pos = Position(
            Vec3d(minecraft.player.lastTickPosX, minecraft.player.lastTickPosY, minecraft.player.lastTickPosZ), Color(Color.HSBtoRGB(colourHue / 360f, 1f, 1f))
        )

        if (PlayerUtil.isMoving || minecraft.player.posY != minecraft.player.lastTickPosY) {
            colourHue++
        }

        // Add position
        positions.add(pos)

        // Update positions
        positions.forEach { it.update() }

        // Remove old positions
        positions.removeIf { !it.isAlive && !infinite.value }
    }

    override fun onRender3D() {
        glPushMatrix()

        // GL stuff
        glDisable(GL_TEXTURE_2D)
        glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA)
        glEnable(GL_LINE_SMOOTH)
        glEnable(GL_BLEND)
        glDisable(GL_DEPTH_TEST)
        glLineWidth(lineWidth.value)

        // Disable lighting
        minecraft.entityRenderer.disableLightmap()

        glBegin(GL_LINE_STRIP)
        for (pos in positions) {
            val renderPosX = minecraft.renderManager.viewerPosX
            val renderPosY = minecraft.renderManager.viewerPosY
            val renderPosZ = minecraft.renderManager.viewerPosZ

            setColour(if (rainbow.value) pos.colour.rgb else colour.value.rgb)

            glVertex3d(pos.position.x - renderPosX, pos.position.y - renderPosY, pos.position.z - renderPosZ)
        }

        // Reset colour
        glColor4d(1.0, 1.0, 1.0, 1.0)

        // End GL
        glEnd()
        glEnable(GL_DEPTH_TEST)
        glDisable(GL_LINE_SMOOTH)
        glDisable(GL_BLEND)
        glEnable(GL_TEXTURE_2D)
        glPopMatrix()
    }

    internal class Position(val position: Vec3d, val colour: Color) {

        // Position's lifespan
        private var lifespan = lifespanValue.value.toLong()

        /**
         * Decreases the lifespan of the position
         */
        fun update() {
            lifespan--
        }

        /**
         * Checks if the position is alive
         *
         * @return If the position is alive
         */
        val isAlive: Boolean
            get() = lifespan > 0
    }

}